projects/output-and-error-handling/main.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+const maxWaitSeconds = 5
+const weatherURL = "http://localhost:8080"
+
+var firstTry = true
+
+func main() {
+	res, err := getWeather()
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v", err)
+		os.Exit(1)
+	}
+	fmt.Println(res)
+}
+
+func getWeather() (string, error) {
+	for {
+
+		res, err := http.Get(weatherURL)
+
+		if err != nil {
+			return "", fmt.Errorf("failed to fetch weather: %v", err)
+		}
+
+		defer res.Body.Close()
+
+		if res.StatusCode == http.StatusOK {
+			body, err := io.ReadAll(res.Body)
+			if err != nil {
+				return "", fmt.Errorf("failed to read response: %v", err)
+			}
+			return string(body), nil
+		}
+
+		if res.StatusCode == http.StatusTooManyRequests {
+			retryAfter := res.Header.Get("Retry-After")
+			seconds, err := parseRetryAfter(retryAfter)
+
+			if err != nil {
+				if retryAfter == "a while" {
+					if firstTry {
+						firstTry = false
+						fmt.Fprintf(os.Stderr, "received 429 with and unknown retry time. will try to retrying in %d seconds...\n", maxWaitSeconds)
+						time.Sleep(time.Duration(maxWaitSeconds) * time.Second)
+						continue
+					} else {
+						return "", fmt.Errorf("received 429 with and unknown retry time again, try again later")
+					}
+				}
+				return "", fmt.Errorf("received 429 but Retry-After header is invalid: %q", retryAfter)
+			}
+
+			if seconds > maxWaitSeconds {
+				return "", fmt.Errorf("server is too busy, retry-after time is too long (%d seconds)", seconds)
+			}
+
+			if seconds == 1 {
+				continue
```

Comment:
Why do we treat 1 second specially? Why is this the correct behaviour?

Reply:
What I'm reading the code doing here is _not_ sleeping at all if it needs to sleep for 1 second? But sleeping for 2 seconds if the time is 2 seconds? Is that right?